text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+
+## Signal lifecycle
+
+OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.
+
+![drawing](img/0143_api_lifecycle.png)
+
+### Lifecycle stages
+
+**Experimental –** Bugs, breaking changes and performance issues may occur.
+
+**Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.
+
+**Deprecated –** this signal has been replaced but is still supported under LTS.
+
+**Removed -** a deprecated signal is no longer supported, and is removed.
+
+All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.
+
+### Stability
+
+Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.
+
+**API Stability -**
+No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.
+
+**SDK Stability -**
+Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
+
+**Semantic Conventions Stability -**
+Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed.
+
+**Contrib Stability -**
+Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.
+
+Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.
+
+Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard.
+
+### Deprecation
+
+In theory, signals could be replaced. When this happens, they are marked as deprecated.
+
+Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.
+
+### Removal
+
+Packages are end-of-life’d by being removed, along with a major version bump.
+
+We currently have no plans for deprecating signals or creating a major version past v1.0.
```

Comment:
What sort of criteria would require a major version past v1.0?  I think in a previous meeting we discussed an additional signal beyond tracing, metrics, and logging?

Reply:
Right now, there are no plans for moving past v1. 

I backed away from additive changes like new signals or stable releases creating a major version bump, as that broke too many assumptions. Only the removal of existing code which requires backwards compatibility would create a major version bump. 

In theory, we could develop a "v2.0" of some kind which deprecated enough existing code that we would want to eventually remove the old code.

Reply:
Wow, are we really thinking that we are creating the perfect OpenTelemetry with v1.0? And we don't plan for v2.0? I understand that we want 1.0 to be solid and long-lived, but saying that it will be the only and final version is... too optimistic?

Reply:
I don't think this is saying there won't be 2.x versions. It just says that we don't have any plans to release breaking changes at this point. We very well may have them later, for sure. But, not *currently*.

Reply:
What @jkwatson said. @iNikem we have a situation where our users are going to rely on v1.0 for years. So, while we can add new things – including a better, totally different tracing API – we can't remove the old things or cease to support them. So that "totally different tracing" API would have to be an entirely new signal which coexists and interoperates with the current tracing signal. v2.0 would mark the end of support for current tracing, not the addition of new awesome tracing. And we don't want to ever end that support, if we could help it.

And to be clear if you think this is crazy – consider that we are already doing exactly this with the OpenTracing API. We invented a new tracing API but continue to support the old one.

Reply:
Well, if you put it that way, then it makes more sense :)

Reply:
So what about v 1.x, are we saying that there will never be a next "part" release?
Arguing that we are only ever going to create a ""totally different tracing" (which wraps the previous one) I believe is missing the reality that we WILL find incremental changes that can be safely added without breaking backward compatibility...

Reply:
Yes of course. Minor version bumps are fine, provided they do not break backwards compatibility.

Reply:
Can we be a bit more specific with minor version bumps. For example:
- **v1.0** - apps compiled against **v1.0** are also API- and ABI-compatible with anything **v1.x**
- **v1.1** - may contain additive changes, e.g. new API (_as in new API in a separate new class for C++_) , considering that any new additions are not breaking ABI compatibility.

Re. **Logging API** specifically - assuming that the Logging is NOT part of GA for **v1.0** , is Logging gonna be part of **v2.0** or **v1.1** ? With a provision of additive changes being OK, assuming these are not ABI-breaking, we'd be able to release Logging in **v1.1** rather than bumping up the major version.

What is your opinion on this?

Reply:
Correct, adding new signals is a minor version bump. So metrics is a minor bump, and logging is a minor bump. All additive changes are a minor version bump.